Add support for streaming responses for Amazon Bedrock LLM

Benyuel opened the pull request:
<!--
Thank you for contributing to LangChainJS! Your PR will appear in our next release under the title you set above. Please make sure it highlights your valuable contribution.

To help streamline the review process, please make sure you read our contribution guidelines:
https://github.com/hwchase17/langchainjs/blob/main/CONTRIBUTING.md

If you are adding an integration (e.g. a new LLM, vector store, or memory), please also read our additional guidelines for integrations:
https://github.com/hwchase17/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

Replace this block with a description of the change, the issue it fixes (if applicable), and relevant context.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle below!
-->

<!-- Remove if not applicable -->

This PR adds support for using the Amazon Bedrock LLM with streaming.  See the added test for an example use.  

Fixes https://github.com/hwchase17/langchainjs/issues/2321

jacoblee93 commented:
This looks really fantastic, but I don't think it really fits as a chat model (there's no dialog turns/messages, for example). The `OpenAIChat` class was really just a one-off when chat models were new.

We can just add streaming to the LLM version I think?

jacoblee93 commented on langchain/src/llms/bedrock-chat.ts:
```diff
@@ -0,0 +1,267 @@
+import { SignatureV4 } from "@aws-sdk/signature-v4";
+import { defaultProvider } from "@aws-sdk/credential-provider-node";
```
Let's combine this into the original `llms/bedrock.ts` and call it adding streaming support!

Benyuel replied:
ok, sounds good.  I will work towards that now.  Thank you for the review

jacoblee93 commented on langchain/src/llms/bedrock-chat.ts:
```diff
@@ -0,0 +1,267 @@
+
+  fetchFn: typeof fetch;
+
+  marshaller: EventStreamMarshaller = new EventStreamMarshaller(
```
I think this has been deprecated?
<img width="876" alt="Screenshot 2023-08-22 at 7 53 14 PM" src="https://github.com/hwchase17/langchainjs/assets/6952323/21fec63c-a275-4309-b3c4-d9d52a7d4774">

Benyuel replied:
Yes, sorry I missed that.  I will replace with a suitable alternative for this 👍

jacoblee93 commented on langchain/src/llms/bedrock-chat.ts:
```diff
@@ -0,0 +1,267 @@
+      );
+    }
+
+    if (this.streaming) {
```
We now have a new interface for `.stream()`:

https://js.langchain.com/docs/guides/expression_language/interface

Also, it'd be more elegant to always use the streaming call and implement non-streaming requests as just concatenating the entire stream:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/llms/ollama.ts#L124

jacoblee93 commented:
Let me know if you need help using the newer package! I'm happy to dig in tomorrow.

jacoblee93 commented:
Thanks for the quick updates!

To get `.stream()` working you'll actually need to implement `_streamResponseChunks` as shown here:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/llms/openai.ts#L452

I can do this later today though - it should be very similar to what you have, and we can share the streaming logic.

jacoblee93 commented:
@Benyuel Refactored a bit so we can support the `.stream()` method and added an integration test - can you confirm `yarn test:single langchain/src/llms/tests/bedrock.int.test.ts` works for you?

I also added retries and the ability to pass an abort signal to cancel in flight requests (hackily).

Also CC @vitaly-ps for a second look?

Benyuel commented:
Thanks so much @jacoblee93!  

I made a minor change to get the tests working for me, and also added passing in `temperature` and `maxTokens` into the request.  

I think this is ready from my perspective

> I also added retries and the ability to pass an abort signal to cancel in flight requests

I tested this works as well 👍

jacoblee93 commented on langchain/src/llms/tests/bedrock.int.test.ts:
```diff
@@ -249,7 +278,7 @@ test("Test Bedrock LLM: stream method", async () => {
     chunks.push(chunk);
   }
-  expect(chunks.length).toBeGreaterThan(1);
+  expect(chunks.length).toBeGreaterThanOrEqual(1);
```
Shouldn't this be greater than 1?

Benyuel replied:
The mock I added does streaming, but just 1 chunk.  
https://github.com/hwchase17/langchainjs/blob/7f2271769488152259791226d4c3767f60919ae9/langchain/src/llms/tests/bedrock.int.test.ts#L321-L337
That's why I swapped this.

jacoblee93 commented on langchain/src/llms/tests/bedrock.int.test.ts:
```diff
@@ -240,6 +242,33 @@ test("Test Bedrock LLM: stream method", async () => {
     maxTokens: 20,
     region,
     model,
+    async fetchFn(
```
Ah I intended this one to test that this works with a live Bedrock instance, without a fetch function mock. Can you try that out?

Benyuel replied:
oh sorry, you can revert the mock I added.  I just assumed no connection to bedrock so mocked it.  

To answer your question though, yes I tried this `.stream` against a live bedrock endpoint too.

jacoblee93 commented:
Hey @Benyuel, I think you changed the test I added to use a mocked function - does this work with a live Bedrock instance?

jacoblee93 commented on langchain/src/llms/bedrock.ts:
```diff
@@ -50,9 +61,9 @@ class BedrockLLMInputOutputAdapter {
       return responseBody.completion;
     } else if (provider === "ai21") {
-      return responseBody.completions[0].data.text;
+      return responseBody.data.text;
```
This was wrong before?

Benyuel replied:
This was correct before for the non-streaming bedrock api.  But when I change to the bedrock `invoke-with-response-stream` endpoint, these change to what's in this PR currently.

jacoblee93 commented:
Looks great! Thanks again for your patience!

uzaymacar commented:
Hey both, I've found this PR from one of the linked issues -- I'm still confused about how to implement streaming with Bedrock similar to what is shown in the docs: https://js.langchain.com/docs/modules/model_io/models/chat/how_to/streaming. Simply changing `ChatOpenAI` with `ChatBedrock` does not work, but maybe I need to do something else?

jacoblee93 commented:
Ah apologies, we really need to update that - instead use:

https://js.langchain.com/docs/expression_language/interface#stream

uzaymacar commented:
Awesome thanks @jacoblee93, I'm now able to stream with Bedrock too 👍
